### PR TITLE
1st part of llext #8180

### PR DESCRIPTION
**Pull request by lyakh**
7 first commits from #8180 not requiring a Zephyr update

### CODE REVIEW

**Comment by marc-hb** [comment]
Thanks for this smaller PR. Still a draft?

File: src/include/sof/lib_manager.h
Comment (kv2019i) [inline src/include/sof/lib_manager.h:86]: This is first time this define is used in basefw build (this is defined when building modules with lmdk but not set for base fw build). I think this has to be a Kconfig for the main build to select loadable module behaviour and/or support both.

**Review by kv2019i — COMMENTED** [review]
Lot of good changes, thanks for splitting these out. One question about the config define.

File: src/include/sof/lib_manager.h
Comment (lyakh) [reply src/include/sof/lib_manager.h:86]: @kv2019i right, yes, I agree that ideally we should be able to build firmware that can load all kinds of modules: IADK, native system-services, and native dynamic. But initially I'd integrate all three approaches into the firmware and then check how we can identify modules at run-time

Comment (lyakh) [reply src/include/sof/lib_manager.h:86]: in fact indeed it isn't correct. `-D__SOF_MODULE_SERVICE_BUILD__=1` is only used to build the **module** itself, not the complete firmware. Converted to draft for a fix

File: src/library_manager/lib_manager.c
Comment (kv2019i) [inline src/library_manager/lib_manager.c:52]: Minor: The magic 3 sticks out a but, some SOF_MAN_SEGMENT_COUNT ?

**Review by kv2019i — APPROVED** [review]
Thanks, looks good now!

File: src/library_manager/lib_manager.c
Comment (jxstelter) [inline src/library_manager/lib_manager.c:207]: Page alignment for sections is caused by secirity reasons. TEXT should be read_only while BSS read/write. Additionally if we want to introduce module level access rights, the PAGE size is minimal quantity we can control.

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:52]: @kv2019i well, here's how they're defined https://github.com/thesofproject/sof/blob/d324af52bd3ceaf502afa15bef12023af595491e/tools/rimage/src/include/rimage/sof/user/manifest.h#L47-L51 and it is the same as https://github.com/thesofproject/sof/blob/d324af52bd3ceaf502afa15bef12023af595491e/tools/rimage/src/include/rimage/sof/user/manifest.h#L103 so, yes, we could add a separate commit to define them and use them in both those structures, but that would modify manifest.h. Even though it wouldn't *actually* change anything, still that probably would be considered an ABI change? Or now?

File: src/library_manager/lib_manager.c
Comment (softwarecki) [inline src/library_manager/lib_manager.c:346]: I would like move this if to `lib_manager_allocate_module_instance`

File: src/library_manager/lib_manager.c
Comment (softwarecki) [inline src/library_manager/lib_manager.c:211]: Lack of support for readonly segment and empty data segment.

**Review by softwarecki — CHANGES_REQUESTED** [review]

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:147]: @softwarecki It should be freed, yes, but I just moved the code from below. And I think some other PRs will rewrite this anyway. So I just moved it out of the way for now to let those other PRs work on this.

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:211]: @softwarecki sorry, is this something that this commit changes? That's how it is now, isn't it?

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:346]: @softwarecki I guess this could be done, and it wouldn't affect the LLEXT flow anyway, but I think it's more robust to have final sizes in this table and to be able to use them with no further modifications wherever needed?

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:217]: @softwarecki I'd rather not do that. As you see there's a difference between .text and .data - we need to additionally invalidate instruction cache for .text. If I only move this write-back call into the mapping function, it will become less logical - for .text the writeback will be inside the function and the icache invalidation will be outside

**Comment by lyakh** [comment]
CI: a suspend-resume test failed on TGL and a pause-resume-capture failed on MTL, none should be related

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:147]: ...and yes, I meant #8544 by @softwarecki

File: src/library_manager/lib_manager.c
Comment (softwarecki) [inline src/library_manager/lib_manager.c:66]: superfluous double checks. Why not just `if (!ctx) return NULL;`?

**Comment by pjdobrowolski** [comment]
I have a mixed feeling about that llext. I only can guess that llext means that it is loadable libraries external however I didn't see any draft of documentation whatsoever. Changes themselves are cosmetic or irrelevant from IADK/Native loadable modules perspective and I would like to review that PR properly. For me without any documentation it is no-go.

File: src/library_manager/lib_manager.c
Comment (softwarecki) [reply src/library_manager/lib_manager.c:147]: In that case, let's not modify the code that will be changed by another PR. There is no point in introducing a merge conflicts in another PR.

File: src/library_manager/lib_manager.c
Comment (softwarecki) [reply src/library_manager/lib_manager.c:217]: Why do you moved a call to `dcache_writeback_region` from the `lib_manager_load_data_from_storage` to this function? i'm ok with icache.

File: src/library_manager/lib_manager.c
Comment (pjdobrowolski) [reply src/library_manager/lib_manager.c:346]: Is ctx->segment_size calculated value for module or is it size of loaded binary? Can you clarify it?

**Review by pjdobrowolski — DISMISSED** [review]

File: src/library_manager/lib_manager.c
Comment (softwarecki) [reply src/library_manager/lib_manager.c:346]: I would prefer all bss-related calculations to be in one place. In my opinion, the name and location (in ctx) of this array suggests storing values read from the manifest. While reviewing the changes in the `lib_manager_allocate_module_instance` function, I was surprised that the division by the number of instances disappeared.

**Review by softwarecki — CHANGES_REQUESTED** [review]

File: src/library_manager/lib_manager.c
Comment (softwarecki) [reply src/library_manager/lib_manager.c:147]: In that case, let's not modify the code that will be changed by another PR. There is no point in introducing a merge conflicts in another PR.

Comment (softwarecki) [reply src/library_manager/lib_manager.c:147]: In that case, let's not modify the code that will be changed by another PR. There is no point in introducing a merge conflicts in another PR.

**Comment by lyakh** [comment]
> I have a mixed feeling about that llext. I only can guess that llext means that it is loadable libraries external however I didn't see any draft of documentation whatsoever. Changes themselves are cosmetic or irrelevant from IADK/Native loadable modules perspective and I would like to review that PR properly. For me without any documentation it is no-go.

https://docs.zephyrproject.org/latest/services/llext/index.html

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:147]: @softwarecki I cannot and shall not sit and wait because I know that somebody else's work is in progress. If your PR gets merged first - I'll update mine, if mine is merged first - you'll update yours

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:66]: @softwarecki I wouldn't really call this superfluous. An alternative would look like
```
if (!ctx || !ctx->desc)
	return NULL;
uint8_t *buffptr = ctx->desc;
```
so you anyway have to check up to two conditions - just as well as now. I'd say it's a matter of preference

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:346]: > Is ctx->segment_size calculated value for module or is it size of loaded binary? Can you clarify it?

@pjdobrowolski that should be rather obvious from the code, sorry. It's calculated from section sizes of the loaded binary.

Comment (lyakh) [reply src/library_manager/lib_manager.c:346]: > I would prefer all bss-related calculations to be in one place. In my opinion, the name and location (in ctx) of this array suggests storing values read from the manifest. While reviewing the changes in the `lib_manager_allocate_module_instance` function, I was surprised that the division by the number of instances disappeared.

@softwarecki currently that division is performed in two locations: in `lib_manager_allocate_module_instance()` and in `lib_manager_free_module_instance()`, after this change it's done only in one location, I think it's better.

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:217]: @softwarecki to make it logical - `lib_manager_load_data_from_storage()` allocates memory and copies whatever data it's asked to handle into it, it doesn't know cache requirements of that data, that is handled by the caller, because the caller knows what data that is. Alternatively we'd need to add some kind of flags to the function to let it know what cache handling it has to perform

File: tools/rimage/src/include/rimage/sof/user/manifest.h
Comment (plbossart) [inline tools/rimage/src/include/rimage/sof/user/manifest.h:28]: maybe explain what LL and EXT stand for? it's not obvious to me. Low Latency, Loadable Library, External, Extension, Exploding,  who knows?

Acronyms are evil.

File: src/library_manager/lib_manager.c
Comment (plbossart) [inline src/library_manager/lib_manager.c:161]: @ujfalusi didn't we make the assumption that a library contains ONE module?

IIRC the firmwares are identified with a UUID, not sure how the library is identified in Linux

File: src/library_manager/lib_manager.c
Comment (plbossart) [inline src/library_manager/lib_manager.c:141]: why the size_t for an index incremented by one?

File: src/library_manager/lib_manager.c
Comment (plbossart) [inline src/library_manager/lib_manager.c]: could be a separate patch

**Review by plbossart — COMMENTED** [review]
Minor comments from a completely clueless reviewer...

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:141]: @plbossart no idea, not my code, moved it as is from the current implementation

File: src/library_manager/lib_manager.c
Comment (lyakh) [reply src/library_manager/lib_manager.c:161]: @plbossart in an earlier version I removed this whole "lib_code" support because, well, it's unused so far and it's unclear if and when and how it would be used. I got complains. So, I keep it as is, just move out to functions

**Comment by pjdobrowolski** [comment]
> > I have a mixed feeling about that llext. I only can guess that llext means that it is loadable libraries external however I didn't see any draft of documentation whatsoever. Changes themselves are cosmetic or irrelevant from IADK/Native loadable modules perspective and I would like to review that PR properly. For me without any documentation it is no-go.
> 
> https://docs.zephyrproject.org/latest/services/llext/index.html

It is just explanation what is llext, there is no sofdocs of its implementation in sof.

File: src/library_manager/lib_manager.c
Comment (pjdobrowolski) [inline src/library_manager/lib_manager.c:624]: lib_manager.c is slowly hitting 1k lines of code. It is hard to understand what and where is jumping, especially when new functions were introduce. Is it a good time to split lib_manager.c for more readable format?

**Comment by lyakh** [comment]
> > > I have a mixed feeling about that llext. I only can guess that llext means that it is loadable libraries external however I didn't see any draft of documentation whatsoever. Changes themselves are cosmetic or irrelevant from IADK/Native loadable modules perspective and I would like to review that PR properly. For me without any documentation it is no-go.
> > 
> > 
> > https://docs.zephyrproject.org/latest/services/llext/index.html
> 
> It is just explanation what is llext, there is no sofdocs of its implementation in sof.

@pjdobrowolski SOF uses many Zephyr APIs: threads, mutexes, work queues, timers, etc., we don't have sofdocs for each of them. But sure, we can add it as a paragraph in SOF loadable module implementation. In short - we use LLEXT for ELF parsing and linking. Also keep in mind, that this PR is just a preparation, it doesn't use LLEXT yet

File: src/library_manager/lib_manager.c
Comment (ujfalusi) [reply src/library_manager/lib_manager.c:161]: No, the kernel side does not assume anything, it is prepared for the module bundles. The kernel tracks the library itself and then checks the module(s) inside:
b700672e2250 ("ASoC: SOF: Intel/IPC4: Support for external firmware libraries")

The lib_code does not have UUID, but in case a library have multiple modules then there should be symlinks:
```
module_bundle.bin
<UUID1>.bin -> module_bundle.bin
<UUID2>.bin -> module_bundle.bin
<UUID3>.bin -> module_bundle.bin
```
The kernel will load a library for a specific module UUID, so multiple UUID pointing to the same library will load that library with all of it's modules once.

File: tools/rimage/src/include/rimage/sof/user/manifest.h
Comment (ujfalusi) [reply tools/rimage/src/include/rimage/sof/user/manifest.h:28]: this conversation made me curious, what is llext stands for? ;) @lyakh

Comment (marc-hb) [reply tools/rimage/src/include/rimage/sof/user/manifest.h:28]: https://www.bing.com/search?q=zephyr+llext

**Comment by lyakh** [comment]
New version. I'm forking lib_manager.c for LLEXT. Sharing it with other loadable module implementations didn't work out. @pjdobrowolski @softwarecki you're personally cordially invited to review this version.

File: tools/rimage/src/include/rimage/sof/user/manifest.h
Comment (kv2019i) [reply tools/rimage/src/include/rimage/sof/user/manifest.h:28]: But it's a fair point this file has nothing to do with Zephyr, so it's not so obvious to figure out this is a term coming from Zephyr.

File: src/library_manager/llext_manager.c
Comment (kv2019i) [inline src/library_manager/llext_manager.c:9]: So this could be more explict and mention "Dynamic module loading functions using Zephyr Linkable Loadable Extensions (LLEXT) interface".

**Review by kv2019i — APPROVED** [review]
Some minor things, but nothing blocking merge. I do wonder how we share https://github.com/thesofproject/sof/pull/8741 -- that's needed here as well, right?

**Comment by lyakh** [comment]
> Some minor things, but nothing blocking merge. I do wonder how we share #8741 -- that's needed here as well, right?

@kv2019i it's not *needed* for this to work, it's needed for SOF to make sure we check what we run, right? And yes, it should work with this PR too, since auth is added to the IPC copying path, which is shared with system-services

File: src/library_manager/llext_manager.c
Comment (lyakh) [reply src/library_manager/llext_manager.c:9]: @kv2019i I propose to address any cosmetic improvements in a follow-up PR, will do one first thing after this is merged

**Comment by lyakh** [comment]
CI failures: a single known multiple-pause-resume failure https://sof-ci.01.org/sofpr/PR8759/build2309/devicetest/index.html?model=TGLU_RVP_NOCODEC-ipc4&testcase=multiple-pause-resume-50
zmain builds fail due to a Zephyr "main" regression
checkpatch complains about `extern` in a C file and a missing SPDX header in a .toml.h, both can be ignored.

**Comment by kv2019i** [comment]
We got the auth bit merged, and now this can go it. @pjdobrowolski @softwarecki you didn't respond the pings this week but Guennadi has now addressed the main concerns and the code is now in separate file, so I assume we are good now. The documentation bit definitely needs to be supported but I think the Zephyr llext docs plus #8180 gives a pretty good view of what this is -- Linux-kernel style loadbale modules support using Zephyr llext mechanism. This won't be for all purposes, but is one option to build modules.

**Pull request merged**
